File: app/src/test/java/blue/koenig/kingsfinances/TestSubject.java
package blue.koenig.kingsfinances;

import java.util.ArrayList;
import java.util.List;

import blue.koenig.kingsfinances.model.calculation.ItemSubject;
import blue.koenig.kingsfinances.model.database.Table;

/**
 * Created by Thomas on 02.01.2018.
 */

public class TestSubject<T> implements ItemSubject<T> {
    private List<Table.OnDeleteListener> deleteListeners = new ArrayList<>();
    private List<Table.OnUpdateListener> updateListeners = new ArrayList<>();
    private List<Table.OnAddListener> addListeners = new ArrayList<>();

    public void add(T item) {
        for (Table.OnAddListener addListener : addListeners) {
            addListener.onAdd(item);
        }
    }

    public void update(T old, T newItem) {
        for (Table.OnUpdateListener updateListener : updateListeners) {
            updateListener.onUpdate(old, newItem);
        }
    }

    public void delete(T item) {
        for (Table.OnDeleteListener onDeleteListener : deleteListeners) {
            onDeleteListener.onDelete(item);
        }
    }

    @Override
    public void addDeleteListener(Table.OnDeleteListener<T> listener) {
        deleteListeners.add(listener);
    }

    @Override
    public void addUpdateListener(Table.OnUpdateListener<T> listener) {
        updateListeners.add(listener);
    }

    @Override
    public void addAddListener(Table.OnAddListener<T> listener) {
        addListeners.add(listener);
    }
}

File: app/src/test/java/blue/koenig/kingsfinances/TestSubject.kt
package blue.koenig.kingsfinances

import com.koenig.commonModel.database.DatabaseItemTable
import com.koenig.commonModel.finance.statistics.ItemSubject
import java.util.*

/**
 * Created by Thomas on 02.01.2018.
 */

open class TestSubject<T> : ItemSubject<T> {
    private val deleteListeners = ArrayList<DatabaseItemTable.OnDeleteListener<T>>()
    private val updateListeners = ArrayList<DatabaseItemTable.OnUpdateListener<T>>()
    private val addListeners = ArrayList<DatabaseItemTable.OnAddListener<T>>()


    fun add(item: T) {
        for (addListener in addListeners) {
            addListener.onAdd(item)
        }
    }

    fun update(old: T, newItem: T) {
        for (updateListener in updateListeners) {
            updateListener.onUpdate(old, newItem)
        }
    }

    fun delete(item: T) {
        for (onDeleteListener in deleteListeners) {
            onDeleteListener.onDelete(item)
        }
    }

    override fun addDeleteListener(listener: DatabaseItemTable.OnDeleteListener<T>) {
        deleteListeners.add(listener)
    }

    override fun addUpdateListener(listener: DatabaseItemTable.OnUpdateListener<T>) {
        updateListeners.add(listener)
    }

    override fun addAddListener(listener: DatabaseItemTable.OnAddListener<T>) {
        addListeners.add(listener)
    }
}
